File: LeetCode/medium/Combination_Sum.cxx
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        std::sort(candidates.begin(), candidates.end());
        
        std::vector<std::vector<int>> found;
        
        std::stack<int> r;
        r.push(target);
        std::stack<std::vector<int>> path;
        path.push(std::vector<int>());
        while (r.size() > 0)
        {
            int reste = r.top();
            r.pop();
            auto p_init = path.top();
            path.pop();
            auto end = std::upper_bound(candidates.begin(), candidates.end(), reste);
            for (auto it = candidates.begin(); it != end; ++it)
            {
                auto p = p_init;
                p.push_back(*it);
                if (reste - *it == 0)
                {
                    found.push_back(p);
                }
                else
                {
                    r.push(reste - *it);
                    path.push(p);
                }
            }
            
        }
        
        std::set<std::vector<int>> ok;
        for (int i = 0; i < found.size(); ++i)
        {
            std::sort(found[i].begin(), found[i].end());
            ok.insert(found[i]);            
        }
        
        std::vector<std::vector<int>> ret;
        for (auto const& v : ok)
        {
            ret.push_back(v);
        }
        
        return ret;
    }
};

File: LeetCode/medium/Combination_Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # def hash(l):
        #     a = [0] * 41
        #     for x in l:
        #         a[x] += 1
        #     return "-".join(map(str, a))
        # rec = set()
        # res = []
        # def solve(cur, r):
        #     h = hash(cur)
        #     if h in rec:
        #         return
        #     rec.add(h)
        #     if r == 0:
        #         res.append(cur)
        #     elif r > 0:
        #         for x in candidates:
        #             solve(cur + [x], r-x)
        # solve([], target)
        # return res
        candidates = sorted(candidates)
        res = []
        def solve(i, cur, cur_sum):
            if i >= len(candidates):
                return
            if cur_sum + candidates[i] == target:
                res.append(cur + [candidates[i]])
            elif cur_sum + candidates[i] < target:
                solve(i, cur + [candidates[i]], cur_sum + candidates[i])
                solve(i+1, cur, cur_sum)
            else:
                solve(i+1, cur, cur_sum)
        solve(0, [], 0)
        return res

